File: KASIKORN Business-Technology Group/Assignment1/Ex5/Ex5.py
import pandas as pd
import random
from datetime import datetime, timedelta
import numpy as np
import time

class Data_In_Loop:
    # V1: Functions inside the loop
    
    def __init__(self):
        self.current_id = 0
        self.setup_master_data()
    
    def setup_master_data(self):
        self.province_codes = ['BKK', 'CNX', 'NTB', 'PTT', 'NKP']
        self.country_codes = ['TH', 'US', 'GB', 'JP', 'SG']
        self.bank_codes = ['79601001', '79601002', '79601003']
        self.positions = ['Manager', 'Staff', 'Director', 'Supervisor']
        self.company_sizes = ['S', 'M', 'L']
        self.first_names = ['Somchai', 'Arthit', 'Kittisak', 'Naphat', 'Mali']
        self.last_names = ['Kasikorn', 'KBTG', 'K+', 'KhunThong', 'MeowJot']
        self.company_types = ['Trading', 'Technology', 'Services']
    
    def generate_thai_id(self):
        self.current_id += 1
        prefix = random.randint(1, 9)
        middle = str(self.current_id).zfill(9)
        check_digit = random.randint(0, 9)
        return f"{prefix}{middle}{check_digit}"
    
    def generate_person_name(self):
        return f"{random.choice(self.first_names)} {random.choice(self.last_names)}"
    
    def generate_company_name(self):
        return f"{random.choice(self.company_types)} {random.choice(['Group', 'Corp', 'Co', 'PLC'])} {self.current_id}"
    
    def generate_data_file(self, num_records, output_file):
        records = []
        num_corporate = int(num_records * 0.05)
        num_individual = num_records - num_corporate
        
        # Generate records with function calls inside the loop
        for _ in range(num_corporate + num_individual):
            is_corporate = len(records) < num_corporate
            
            if is_corporate:
                company_name = self.generate_company_name()
                record = {
                    'CUSTID': f"C{self.generate_thai_id()}",
                    'FULL_NAME': company_name,
                    'CUST_TYPE': 'P01',
                    'IDENTITY_NUM': self.generate_thai_id()
                }
            else:
                record = {
                    'CUSTID': f"I{self.generate_thai_id()}",
                    'FULL_NAME': self.generate_person_name(),
                    'CUST_TYPE': 'P02',
                    'IDENTITY_NUM': self.generate_thai_id()
                }
            records.append(record)
            
        df = pd.DataFrame(records)
        df.to_csv(output_file, index=False)
        return len(records)

class Data_Out_Loop:
    # V2: Data outside loop
    
    def __init__(self):
        self.current_id = 0
        self.setup_master_data()
        
    def setup_master_data(self):
        self.province_codes = ['BKK', 'CNX', 'NTB', 'PTT', 'NKP']
        self.country_codes = ['TH', 'US', 'GB', 'JP', 'SG']
        self.bank_codes = ['79601001', '79601002', '79601003']
        self.positions = ['Manager', 'Staff', 'Director', 'Supervisor']
        self.company_sizes = ['S', 'M', 'L']
        
        self.first_names = ['Somchai', 'Arthit', 'Kittisak', 'Naphat', 'Mali']
        self.last_names = ['Kasikorn', 'KBTG', 'K+', 'KhunThong', 'MeowJot']
        self.person_names = [f"{f} {l}" for f in self.first_names for l in self.last_names]
        
        company_types = ['Trading', 'Technology', 'Services']
        company_suffixes = ['Group', 'Corp', 'Co', 'PLC']
        self.company_base_names = [f"{t} {s}" for t in company_types for s in company_suffixes]
        
        self.id_prefixes = list(range(1, 10))
        self.id_check_digits = list(range(10))
        self.id_template = "{prefix}{middle:09d}{check}"
        self.current_id = 0
    
    def generate_batch_ids(self, batch_size):
        """Generate multiple IDs at once"""
        ids = []
        for _ in range(batch_size):
            self.current_id += 1
            ids.append(self.id_template.format(
                prefix=random.choice(self.id_prefixes),
                middle=self.current_id,
                check=random.choice(self.id_check_digits)
            ))
        return ids
    
    def generate_data_file(self, num_records, output_file):
        records = []
        num_corporate = int(num_records * 0.05)
        num_individual = num_records - num_corporate
        
        all_ids = self.generate_batch_ids(num_records * 2)  # Times 2 cause we need ID for both CUSTID and IDENTITY_NUM
        id_index = 0
        
        for i in range(num_corporate):
            company_name = f"{random.choice(self.company_base_names)} {i+1}"
            record = {
                'CUSTID': f"C{all_ids[id_index]}",
                'FULL_NAME': company_name,
                'CUST_TYPE': 'P01',
                'IDENTITY_NUM': all_ids[id_index + 1]
            }
            records.append(record)
            id_index += 2
        
        for _ in range(num_individual):
            record = {
                'CUSTID': f"I{all_ids[id_index]}",
                'FULL_NAME': random.choice(self.person_names),
                'CUST_TYPE': 'P02',
                'IDENTITY_NUM': all_ids[id_index + 1]
            }
            records.append(record)
            id_index += 2
            
        df = pd.DataFrame(records)
        df.to_csv(output_file, index=False)
        return len(records)

def run_performance_test(generator_class, sizes):
    # Run performance test for a given class
    results = []
    generator = generator_class()
    
    for size in sizes:
        start_time = time.time()
        output_file = f"customer_data_{generator_class.__name__}_{size}records.csv"
        
        print(f"\nGenerating {size} records using {generator_class.__name__}...")
        records_generated = generator.generate_data_file(size, output_file)
        
        duration = time.time() - start_time
        memory_usage = df = pd.read_csv(output_file).memory_usage(deep=True).sum() / 1024 / 1024 
        
        result = {
            'size': size,
            'duration': duration,
            'records_per_second': records_generated/duration,
            'memory_mb': memory_usage
        }
        results.append(result)
        
        print(f"Generated {records_generated} records in {duration:.2f} seconds")
        print(f"Speed: {records_generated/duration:.2f} records/second")
        print(f"Memory usage: {memory_usage:.2f} MB")
    
    return results

def main():
    sizes = [10000, 100000, 1000000]
    
    print("Testing function in loop...")
    in_loop_results = run_performance_test(Data_In_Loop, sizes)
    
    print("\nTesting function out loop...")
    optimized_results = run_performance_test(Data_Out_Loop, sizes)
    
    # Compare results
    print("\nPerformance Comparison:")
    print("\nSize\tMethod\t\tDuration\tSpeed\t\tMemory")
    print("-" * 70)
    
    for i, size in enumerate(sizes):
        in_loop = in_loop_results[i]
        opt = optimized_results[i]
        
        print(f"{size:,}\tIn Loop\t\t{in_loop['duration']:.2f}s\t{in_loop['records_per_second']:.0f} r/s\t{in_loop['memory_mb']:.1f}MB")
        print(f"{size:,}\tOptimized\t{opt['duration']:.2f}s\t{opt['records_per_second']:.0f} r/s\t{opt['memory_mb']:.1f}MB")
        print(f"Speedup: {in_loop['duration']/opt['duration']:.2f}x")
        print("-" * 70)

if __name__ == "__main__":
    main()

File: KASIKORN Business-Technology Group/Assignment1/Ex5/Ex5.java
package Assignment1.Ex5;

import java.io.FileWriter;
import java.io.IOException;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class Record {
    String custId;
    String fullName;
    String custType;
    String identityNum;

    Record(String custId, String fullName, String custType, String identityNum) {
        this.custId = custId;
        this.fullName = fullName;
        this.custType = custType;
        this.identityNum = identityNum;
    }

    @Override
    public String toString() {
        return String.format("%s,%s,%s,%s", custId, fullName, custType, identityNum);
    }
}

class DataInLoop {
    private int currentId = 0;
    private final Random random = new Random();
    private final List<String> provinceCodes;
    private final List<String> countryCodes;
    private final List<String> bankCodes;
    private final List<String> positions;
    private final List<String> companySizes;
    private final List<String> firstNames;
    private final List<String> lastNames;
    private final List<String> companyTypes;

    public DataInLoop() {
        this.provinceCodes = Arrays.asList("BKK", "CNX", "NTB", "PTT", "NKP");
        this.countryCodes = Arrays.asList("TH", "US", "GB", "JP", "SG");
        this.bankCodes = Arrays.asList("79601001", "79601002", "79601003");
        this.positions = Arrays.asList("Manager", "Staff", "Director", "Supervisor");
        this.companySizes = Arrays.asList("S", "M", "L");
        this.firstNames = Arrays.asList("Somchai", "Arthit", "Kittisak", "Naphat", "Mali");
        this.lastNames = Arrays.asList("Kasikorn", "KBTG", "K+", "KhunThong", "MeowJot");
        this.companyTypes = Arrays.asList("Trading", "Technology", "Services");
    }

    private String generateThaiId() {
        currentId++;
        int prefix = random.nextInt(9) + 1;
        String middle = String.format("%09d", currentId);
        int checkDigit = random.nextInt(10);
        return prefix + middle + checkDigit;
    }

    private String generatePersonName() {
        return firstNames.get(random.nextInt(firstNames.size())) + " " +
               lastNames.get(random.nextInt(lastNames.size()));
    }

    private String generateCompanyName() {
        String companyType = companyTypes.get(random.nextInt(companyTypes.size()));
        String suffix = Arrays.asList("Group", "Corp", "Co", "PLC").get(random.nextInt(4));
        return String.format("%s %s %d", companyType, suffix, currentId);
    }

    public int generateDataFile(int numRecords, String outputFile) throws IOException {
        List<Record> records = new ArrayList<>();
        int numCorporate = (int)(numRecords * 0.05);
        int numIndividual = numRecords - numCorporate;

        for (int i = 0; i < numRecords; i++) {
            boolean isCorporate = i < numCorporate;
            Record record;

            if (isCorporate) {
                String companyName = generateCompanyName();
                record = new Record(
                    "C" + generateThaiId(),
                    companyName,
                    "P01",
                    generateThaiId()
                );
            } else {
                record = new Record(
                    "I" + generateThaiId(),
                    generatePersonName(),
                    "P02",
                    generateThaiId()
                );
            }
            records.add(record);
        }

        try (FileWriter writer = new FileWriter(outputFile)) {
            writer.write("CUSTID,FULL_NAME,CUST_TYPE,IDENTITY_NUM\n");
            for (Record record : records) {
                writer.write(record.toString() + "\n");
            }
        }

        return records.size();
    }
}

class DataOutLoop {
    private int currentId = 0;
    private final Random random = new Random();
    private final List<String> provinceCodes;
    private final List<String> countryCodes;
    private final List<String> bankCodes;
    private final List<String> positions;
    private final List<String> companySizes;
    private final List<String> personNames;
    private final List<String> companyBaseNames;
    private final List<Integer> idPrefixes;
    private final List<Integer> idCheckDigits;

    public DataOutLoop() {
        this.provinceCodes = Arrays.asList("BKK", "CNX", "NTB", "PTT", "NKP");
        this.countryCodes = Arrays.asList("TH", "US", "GB", "JP", "SG");
        this.bankCodes = Arrays.asList("79601001", "79601002", "79601003");
        this.positions = Arrays.asList("Manager", "Staff", "Director", "Supervisor");
        this.companySizes = Arrays.asList("S", "M", "L");

        List<String> firstNames = Arrays.asList("Somchai", "Arthit", "Kittisak", "Naphat", "Mali");
        List<String> lastNames = Arrays.asList("Kasikorn", "KBTG", "K+", "KhunThong", "MeowJot");
        
        this.personNames = firstNames.stream()
            .flatMap(f -> lastNames.stream().map(l -> f + " " + l))
            .collect(Collectors.toList());

        List<String> companyTypes = Arrays.asList("Trading", "Technology", "Services");
        List<String> companySuffixes = Arrays.asList("Group", "Corp", "Co", "PLC");
        
        this.companyBaseNames = companyTypes.stream()
            .flatMap(t -> companySuffixes.stream().map(s -> t + " " + s))
            .collect(Collectors.toList());

        this.idPrefixes = IntStream.rangeClosed(1, 9).boxed().collect(Collectors.toList());
        this.idCheckDigits = IntStream.range(0, 10).boxed().collect(Collectors.toList());
    }

    private List<String> generateBatchIds(int batchSize) {
        List<String> ids = new ArrayList<>();
        for (int i = 0; i < batchSize; i++) {
            currentId++;
            ids.add(String.format("%d%09d%d",
                idPrefixes.get(random.nextInt(idPrefixes.size())),
                currentId,
                idCheckDigits.get(random.nextInt(idCheckDigits.size()))
            ));
        }
        return ids;
    }

    public int generateDataFile(int numRecords, String outputFile) throws IOException {
        List<Record> records = new ArrayList<>();
        int numCorporate = (int)(numRecords * 0.05);
        int numIndividual = numRecords - numCorporate;

        List<String> allIds = generateBatchIds(numRecords * 2);
        int idIndex = 0;

        for (int i = 0; i < numCorporate; i++) {
            String companyName = companyBaseNames.get(random.nextInt(companyBaseNames.size())) + " " + (i + 1);
            records.add(new Record(
                "C" + allIds.get(idIndex),
                companyName,
                "P01",
                allIds.get(idIndex + 1)
            ));
            idIndex += 2;
        }

        for (int i = 0; i < numIndividual; i++) {
            records.add(new Record(
                "I" + allIds.get(idIndex),
                personNames.get(random.nextInt(personNames.size())),
                "P02",
                allIds.get(idIndex + 1)
            ));
            idIndex += 2;
        }

        try (FileWriter writer = new FileWriter(outputFile)) {
            writer.write("CUSTID,FULL_NAME,CUST_TYPE,IDENTITY_NUM\n");
            for (Record record : records) {
                writer.write(record.toString() + "\n");
            }
        }

        return records.size();
    }
}

class Result {
    int size;
    double duration;
    double recordsPerSecond;
    long memoryMb;

    Result(int size, double duration, double recordsPerSecond, long memoryMb) {
        this.size = size;
        this.duration = duration;
        this.recordsPerSecond = recordsPerSecond;
        this.memoryMb = memoryMb;
    }
}

public class Ex5 {
    private static List<Result> runPerformanceTest(String className, List<Integer> sizes) {
        List<Result> results = new ArrayList<>();
        Object generator = className.equals("DataInLoop") ? new DataInLoop() : new DataOutLoop();

        for (int size : sizes) {
            System.out.printf("\nGenerating %d records using %s...\n", size, className);
            
            long startTime = Instant.now().toEpochMilli();
            String outputFile = String.format("customer_data_%s_%drecords.csv", className, size);
            
            try {
                int recordsGenerated;
                if (generator instanceof DataInLoop) {
                    recordsGenerated = ((DataInLoop)generator).generateDataFile(size, outputFile);
                } else {
                    recordsGenerated = ((DataOutLoop)generator).generateDataFile(size, outputFile);
                }

                double duration = (Instant.now().toEpochMilli() - startTime) / 1000.0;
                long memoryUsage = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
                memoryUsage = memoryUsage / (1024 * 1024); // Convert to MB

                Result result = new Result(
                    size,
                    duration,
                    recordsGenerated / duration,
                    memoryUsage
                );
                results.add(result);

                System.out.printf("Generated %d records in %.2f seconds\n", recordsGenerated, duration);
                System.out.printf("Speed: %.2f records/second\n", recordsGenerated / duration);
                System.out.printf("Memory usage: %d MB\n", memoryUsage);

            } catch (IOException e) {
                System.err.println("Error writing to file: " + e.getMessage());
            }
        }

        return results;
    }

    public static void main(String[] args) {
        List<Integer> sizes = Arrays.asList(10000, 100000, 1000000);

        System.out.println("Testing function in loop...");
        List<Result> inLoopResults = runPerformanceTest("DataInLoop", sizes);

        System.out.println("\nTesting function out loop...");
        List<Result> optimizedResults = runPerformanceTest("DataOutLoop", sizes);

        // Compare results
        System.out.println("\nPerformance Comparison:");
        System.out.println("\nSize\tMethod\t\tDuration\tSpeed\t\tMemory");
        System.out.println("-".repeat(70));

        for (int i = 0; i < sizes.size(); i++) {
            Result inLoop = inLoopResults.get(i);
            Result opt = optimizedResults.get(i);

            System.out.printf("%,d\tIn Loop\t\t%.2fs\t%.0f r/s\t%dMB\n",
                inLoop.size, inLoop.duration, inLoop.recordsPerSecond, inLoop.memoryMb);
            System.out.printf("%,d\tOptimized\t%.2fs\t%.0f r/s\t%dMB\n",
                opt.size, opt.duration, opt.recordsPerSecond, opt.memoryMb);
            System.out.printf("Speedup: %.2fx\n", inLoop.duration / opt.duration);
            System.out.println("-".repeat(70));
        }
    }
}